src/Microsoft.AspNetCore.OData/Extensions/ODataServiceProviderConfigExtenions.cs

```diff
@@ -0,0 +1,240 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.OData.Common;
+using Microsoft.AspNet.OData.Query;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData;
+
+namespace Microsoft.AspNet.OData.Extensions
+{
+    internal static class ODataConfigurationExtensions
+    {
+        public static void SetDefaultQuerySettings(this IServiceProvider serviceProvider, DefaultQuerySettings defaultQuerySettings)
+        {
+            if (serviceProvider == null)
+            {
+                throw Error.ArgumentNull("serviceProvider");
+            }
+
+            if (defaultQuerySettings == null)
+            {
+                throw Error.ArgumentNull("defaultQuerySettings");
+            }
+
+            if (!defaultQuerySettings.MaxTop.HasValue || defaultQuerySettings.MaxTop > 0)
+            {
+                ModelBoundQuerySettings.DefaultModelBoundQuerySettings.MaxTop = defaultQuerySettings.MaxTop;
+            }
+
+            DefaultQuerySettings querySettings = serviceProvider.GetRequiredService<DefaultQuerySettings>();
+            if (querySettings == null)
+            {
+                throw Error.InvalidOperation(SRResources.MissingODataServices, nameof(DefaultQuerySettings));
+            }
+
+            querySettings = defaultQuerySettings;
+        }
```

Comment:
How does this even work?

Reply:
Synced offline, right.

---
In reply to: [375972141](https://github.com/OData/WebApi/pull/2035#discussion_r375972141) [](ancestors = 375972141)